### ISSUE TITLE
openstack port binding failed when using f5 undercloud and network type vxlan

### ISSUE THREAD

**Issue by wynn5a** [F5Networks/f5-openstack-lbaasv2-driver#635]
* Title: openstack port binding failed when using f5 undercloud and network type vxlan
* Attachments: For bugs, attach the agent log and configuration files
  * /etc/neutron/services/f5/f5-openstack-agent.ini
  * /var/log/neutron/f5-openstack-agent.log

   [Attachments](https://www.dropbox.com/s/el281ny0x14fn0b/files.zip?dl=0)

* Details: For bugs, copy and paste the following template into your new issue and fill it out.

#### Agent Version

```
f5-icontrol-rest==1.3.0
f5-openstack-agent==9.3.1
f5-openstack-heat-plugins==9.0.2
f5-openstack-lbaasv2-driver==9.3.1
f5-sdk==2.3.3
```

#### Operating System
Cent OS 7

#### OpenStack Release
Mitaka

#### Bug Severity 
For bugs enter the bug severity level. Do not set any labels.

Severity: 5  --> I'm not sure if this is a bug, and I am looking for help here.

Severity level definitions:
 1. **Severity 1 (Critical)** : Defect is causing systems to be offline and/or nonfunctional. immediate attention is required.
 2. **Severity 2 (High)** : Defect is causing major obstruction of system operations.
 3. **Severity 3 (Medium)** : Defect is causing intermittent errors in system operations.
 4. **Severity 4 (Low)** : Defect is causing infrequent interuptions in system operations.
 5. **Severity 5 (Trival)** : Defect is not causing any interuptions to system operations, but none-the-less is a bug.

#### Description

1. install devstack and deploy openstack with LBaaS v2 enabled

2. install F5 agent and F5 driver follow [document](http://f5-openstack-lbaasv2-driver.readthedocs.io/en/latest/map_f5-lbaasv2-user-guide.html) 

3. do some configuration following douments.

4. create a load-balancer with `neutron lbaas-loadbalancer-create --name lb1 private-subnet`

5. Port binding fail in neutron server log and port status is DOWN.

#### Deployment
all in one openstack deployment
F5 VE standalone

**Comment by mattgreene** [comment]
@ssorenso : Please triage this issue and consult with Paul if necessary because heat-plugin is installed in parallel with lbaas packages.

**Comment by richbrowne** [comment]
This looks like a known issue.  Will be fixed with mech driver.  The VIP
port status DOWN is actually ok, because we create a new port for the self
IP.

**Comment by ssorenso** [comment]
Hi @wynn5a,

My colleague @richbrowne is correct.  For further information, if you do not supply a vip_port, and your subnet starts at 10.0.0.10 as your configuration implies:
```
{   "description": "",
    "admin_state_up": True,
    "network_id": u"2ba3cbd7-c7bf-496c-a1fa-10c524cfa4a4",
    "tenant_id": "945de8b174ac40219588f0f7e2c30e2c",
    "provisioning_status": "PENDING_CREATE",
    "vip_port": {"status": u"DOWN",
        "binding:host_id": u"",
        "description": None,
        "allowed_address_pairs": [],
        "extra_dhcp_opts": [],
        "updated_at": "2017-06-27T06:11:18",
        "device_owner": u"neutron:LOADBALANCERV2",
        "port_security_enabled": True,
        "binding:profile": {},
        "fixed_ips": [
            {"subnet_id": u"d8834d2e-eda8-4879-813c-5d86420ed2e9",
                "ip_address": u"10.0.0.10"
                },
            {"subnet_id": u"9cac3b7a-9831-46c9-b045-75bea5dc53da",
                "ip_address": u"fdea:212f:d14d:0:f816:3eff:fed5:aec4"
                }
            ],
        "id": u"1d815165-4db7-44fe-8ed7-5e3e77712388",
        "security_groups": [u"71736135-4980-42f7-8bf4-136098c5e6ef"],
        "device_id": u"ce6b45f4-51f0-4453-aa57-629badbd2e07",
        "name": u"loadbalancer-ce6b45f4-51f0-4453-aa57-629badbd2e07",
        "admin_state_up": False,
        "network_id": u"2ba3cbd7-c7bf-496c-a1fa-10c524cfa4a4",
        "dns_name": None,
        "binding:vif_details": {},
        "binding:vnic_type": u"normal",
        "binding:vif_type": u"unbound",
        "tenant_id": u"945de8b174ac40219588f0f7e2c30e2c",
        "mac_address": u"fa:16:3e:d5:ae:c4",
        "created_at": "2017-06-27T06:11:18"
        },
    "vip_subnet_id": "d8834d2e-eda8-4879-813c-5d86420ed2e9",
    "listeners": [],
    "vip_address": "10.0.0.10",
    "vip_port_id": "1d815165-4db7-44fe-8ed7-5e3e77712388",
    "provider": None,
    "pools": [],
    "id": "ce6b45f4-51f0-4453-aa57-629badbd2e07",
    "operating_status": "OFFLINE",
    "name": "lb1"
}
```
then the vip_port will auto-resolve via DHCP or static to the following (typically):
+0 - gateway (if any, otherwise to the next)
+1 - DHCP (if any, otherwise "")
+2 - Virtual port NAT direction
+3 - Self IP (vip port)
+4 - SNAT

Leaving 5+ to any members on the subnet.  This is the typical, default resolution.  It also assumes quite a few things about what you're trying to do.  With just knowing that you are going to use a vip_port, this could be assigned to an address of +3 (what it will auto-resolve to) to your starting octet or +5 or above.  I'll speak further with Rich tomorrow to get further details regarding the fix in the mech driver he mentions as it may be that we'll auto-resolve conflicts to a +0 as you appear to be attempting.

Please let me know if you have any further questions in the meantime.

**Comment by wynn5a** [comment]
@ssorenso Thanks for your help.

This is the first time I have tried to integrate BIG-IP and OpenStack together.

When I finished the integration configuration and created a load balance for two vm instances in OpenStack, it can't work. The BIG-IP xui shows my virtual server is down because the children pool member(s) are down.
Log files can be found [here](https://www.dropbox.com/s/nhzlubm4s8mih1z/files-virtual_server_down.zip?dl=0).
Could you please help to check if I am miss something or doing something wrong? Thanks you.

**Comment by ssorenso** [comment]
Hi @wynn5a,

It appears as though your loadbalancer is still in `PENDING_UPDATE` state.  This can be viewed with:
`neutron lbaas-loadbalancer-list`
For all of your loadbalancers.  What i'm trying to find out is if it's in error and stuck in this state or if this is simply updating the changes that you've made.  Looking at the last 2 pollings of the partition's settings, I see that the loadbalancer was still experiencing some level of change.

Would you mind sending me more data?  I don't yet see a root cause as to why the state would remain this way, but I speculate that it's the frequency of changes (as a rough guess).  This is based upon the fact that you made quite a few changes in a 10-20 second span and it may just not have given the runtime long enough to propagate back depending on network performance, system performance, etc.

Thanks!

**Comment by wynn5a** [comment]
@ssorenso Thank you for taking the time to help me.

1. I don't use any script to create the load balance only OpenStack dashboard and neutron-cli.
2. `neutron lbaas-loadbalancer-list` shows the load balancer status is in `ACTIVE`.

All of logs after creating the load balance and my OpenStack test environment dashboard user info can be found [here](https://www.dropbox.com/s/spi6atw3mmc726g/long_log_files.zip?dl=0)

**Comment by wynn5a** [comment]
The network administrator told me that public IP for my OpenStack test env was reclaimed earlier today.
Please let me know if you want more information.

**Comment by ssorenso** [comment]
Hi @wynn5a,

**Situational Note:**
I got a second opinion from @jlongstaf and I will be out on vacation until 7/6.  @jlongstaf has agreed to answer any further questions for you if you have any while I am out.

**Review of the data:**
After reviewing the data that you have provided me today, I have a suggested next plan of action:
```
'loadbalancer': {
    u'description': u'',
    u'provisioning_status': u'PENDING_CREATE', <<< HERE
    u'tenant_id': u'945de8b174ac40219588f0f7e2c30e2c',
    u'admin_state_up': True,
    u'provider': None,
    u'pools': [],
    u'listeners': [],
    u'vip_port_id': u'8f8d258e-1e3e-412b-8f70-0eddc0b4b675',
    u'vip_address': u'10.0.0.16',
    u'vip_subnet_id': u'd8834d2e-eda8-4879-813c-5d86420ed2e9',
    u'id': u'05ef9f12-3449-437c-96b5-a5288b8d887c',
    u'operating_status': u'OFFLINE',
    u'name': u'lb3'
    },
u'service': {
    u'subnets': {
            u'd8834d2e-eda8-4879-813c-5d86420ed2e9': {u'description': u'',
            u'enable_dhcp': True,
            u'network_id': u'2ba3cbd7-c7bf-496c-a1fa-10c524cfa4a4',
            u'tenant_id': u'945de8b174ac40219588f0f7e2c30e2c',
            u'created_at': u'2017-06-26T15:12:18',
            u'dns_nameservers': [],
            u'updated_at': u'2017-06-26T15:12:18',
            u'gateway_ip': u'10.0.0.1',
            u'ipv6_ra_mode': None,
            u'allocation_pools': [{u'start': u'10.0.0.2',
            u'end': u'10.0.0.254'}],
            u'host_routes': [],
            u'shared': False,
            u'ip_version': 4,
            u'ipv6_address_mode': None,
            u'cidr': u'10.0.0.0/24',
            u'id': u'd8834d2e-eda8-4879-813c-5d86420ed2e9',
            u'subnetpool_id': None,
            u'name': u'private-subnet'
        }
    },
    u'listeners': [],
    u'healthmonitors': [],
    u'members': [],
    u'l7policy_rules': [],
    u'pools': [],
    u'l7policies': [],
    u'networks': {
        u'2ba3cbd7-c7bf-496c-a1fa-10c524cfa4a4': {
            u'provider:physical_network': None,
            u'ipv6_address_scope': None,
            u'port_security_enabled': True,
            u'mtu': 1450,
            u'id': u'2ba3cbd7-c7bf-496c-a1fa-10c524cfa4a4',
            u'router:external': False,
            u'availability_zone_hints': [],
            u'availability_zones': [u'nova'],
            u'ipv4_address_scope': None,
            u'shared': False,
            u'status': u'ACTIVE',
            u'subnets': [u'9cac3b7a-9831-46c9-b045-75bea5dc53da',
            u'd8834d2e-eda8-4879-813c-5d86420ed2e9'],
            u'description': u'',
            u'tags': [],
            u'updated_at': u'2017-06-26T15:12:15',
            u'provider:segmentation_id': 1089,
            u'name': u'private',
            u'admin_state_up': True,
            u'tenant_id': u'945de8b174ac40219588f0f7e2c30e2c',
            u'created_at': u'2017-06-26T15:12:15',
            u'provider:network_type': u'vxlan',
            u'vlan_transparent': None
        }
    },
    u'loadbalancer': {
        u'vxlan_vteps': [
            u'10.0.0.250',
            u'125.208.28.74'
        ],
        u'name': u'lb3',
        u'provisioning_status': u'PENDING_CREATE',
        u'network_id': u'2ba3cbd7-c7bf-496c-a1fa-10c524cfa4a4',
        u'tenant_id': u'945de8b174ac40219588f0f7e2c30e2c',
        u'admin_state_up': True,
        u'provider': None,
        u'id': u'05ef9f12-3449-437c-96b5-a5288b8d887c',
        u'gre_vteps': [],
        u'pools': [],
        u'listeners': [],
        u'vip_port_id': u'8f8d258e-1e3e-412b-8f70-0eddc0b4b675',
        u'vip_address': u'10.0.0.16',
        u'vip_subnet_id': u'd8834d2e-eda8-4879-813c-5d86420ed2e9',
        u'vip_port': {
            u'allowed_address_pairs': [],
            u'extra_dhcp_opts': [],
            u'updated_at': u'2017-06-29T09:01:12',
            u'device_owner': u'neutron:LOADBALANCERV2',
            u'port_security_enabled': True,
            u'binding:profile': {},
            u'fixed_ips': [{u'subnet_id': u'd8834d2e-eda8-4879-813c-5d86420ed2e9',
            u'ip_address': u'10.0.0.16'},
            {u'subnet_id': u'9cac3b7a-9831-46c9-b045-75bea5dc53da',
            u'ip_address': u'fdea:212f:d14d:0:f816:3eff:fe7a:7a4f'}],
            u'id': u'8f8d258e-1e3e-412b-8f70-0eddc0b4b675',
            u'security_groups': [u'71736135-4980-42f7-8bf4-136098c5e6ef'],
            u'binding:vif_details': {},
            u'binding:vif_type': u'unbound',
            u'mac_address': u'fa:16:3e:7a:7a:4f',
            u'status': u'DOWN',  << REFERENCED LATER
            u'binding:host_id': u'',
            u'description': None,
            u'device_id': u'05ef9f12-3449-437c-96b5-a5288b8d887c',
            u'name': u'loadbalancer-05ef9f12-3449-437c-96b5-a5288b8d887c',
            u'admin_state_up': False,
            u'network_id': u'2ba3cbd7-c7bf-496c-a1fa-10c524cfa4a4',
            u'dns_name': None,
            u'created_at': u'2017-06-29T09:01:12',
            u'binding:vnic_type': u'normal',
            u'tenant_id': u'945de8b174ac40219588f0f7e2c30e2c'
        },
        u'operating_status': u'OFFLINE',
        u'description': u''
    }
```
So, I cannot determine whether or not the loadbalancer is in a non-PENDING state from the logs.  However, I do see that your `vip_port` is in a `DOWN` state, and I’m guessing this is the root problem.  This is likely the issue as neutron will be attempting to route the traffic through this port, and while it’s in that state, it will not allow your members to come online.

With that said, because I cannot ever see the loadbalancer exit that PENDING state, and with the known condition with neutron might lock out that status field.  This can occur if the commands partitioning child objects onto the loadbalancer in fast succession before the system can provision the children; however, with you stating that it is in an ACTIVE state, this is likely not the problem.

If you have attempted to bring `vip_port` online already, but if failed, then I would suggest temporarily removing it to see if the members come online at the 10.0.0.3 self_ip auto-provisioning.  Changing it then should be as easy as re-creating the `vip_port` and putting it into an ONLINE state.  If this is not possible, then please capture the re-attempt to bring the `vip_port` online and send that to us so that we may further troubleshoot this for us.

**Observations:**

* The `vip_port` assigned for the `self_ip` on the loadbalancer is in a `status` of `DOWN`
  * This, if an attempt were to be put into an UP state would likely cause, at this unstable PENDING state to cause the lbaas `loadbalancer` to enter into an `ERROR` state
  * This is speculation based upon previous observations of neutron
* Based upon the information seen, the first entry visible was for the introduction of the `vip_port`
  * This appears to be the first object provisioned on top of the `loadbalancer`
  * Without logs from prior to this entry, I cannot determine whether or not neutron’s database had freed the `loadbalancer` state from PENDING
    * The fact that the `loadbalancer` is in a PENDING state on this first entry, though, highly suggests that there was a race condition
* The loadbalancer throughout all update requests from neutron have the loadbalancer in a `PENDING` state, and I don’t see it leave this… however…
  * This has been ruled out as you stated that the state is now ACTIVE
  * If this were not the case, then…
    * The loadbalancer’s subsequent objects cannot change state while it is in this state
    * This is the `provisioning_status`
    * Providing this information to explain previous conclusions


**Possible Root Causes:**
_For the members not coming online:_
This is due to the `vip_port` being in a DOWN state.  Here, neutron has partitioned the self_ip for the BIG-IP® to use the allotted address/neutron port.  With it being offline, the BIG-IP is set to be restricted for this port.

**Next Steps (Suggested Action):**
I would suggest attempting to put the `vip_port` into an UP state from the DOWN state it currently is in.  From there, please capture the output if it still fails to come online and get back to us with it and the same logs you’ve provided thus far.

From there, we may have a better understanding of what’s gone wrong, or the issue will be resolved.

Please let us know if there’s any other questions that you might have over the above or over the action plan.  Please forgive me if you’ve already attempted this, but please then provide this output.

Thanks,
ssorenso

**Comment by wynn5a** [comment]
@ssorenso Thank you for your professional advice, I will try as you suggested. 
If there is any progress, I will update it here in time.
Hava a nice holiday!

**Comment by mattgreene** [comment]
Hi @wynn5a, do you have any updates on this issue or can we mark it as closed?

**Comment by wynn5a** [comment]
@mattgreene Sorry for late response. 
I've changed to Heat plugin because nothing happened when trying as @ssorenso suggested.
Close this issue and thanks your help.
